Adds an option to simplify repeatedly instead of just once

hannes-steffenhagen-diffblue opened the pull request:
I did some tests to check if simplifying goto code twice would yield the same result (i.e. simplify once simplifies as much as we can). This turned out not to be the case.

This adds an option (--simplify-full) to simplify the goto code repeatedly until nothing changes anymore, which should address that issue.

hannes-steffenhagen-diffblue committed:
Adds an option to simplify repeatedly instead of just once

hannes-steffenhagen-diffblue commented:
It has been suggested that `simplify_program_once` should return a `bool` to simplify the check in `simplify_program_fully`

tautschnig commented:
The commit description sounds like this is fixing the symptoms rather than fixing the root cause?

hannes-steffenhagen-diffblue committed:
Allows simplification of expressions containing non-const

Previously, abstract_objectt::expression_transform returned top
whenever one of the subexpressions of an operator was non-const.

This change allows it to continue and try to replace the whole
expression instead (e.g. in cases like `(x-x)*nondet_int()`).

This also necessitated a change in abstract_environmentt::assign;
Previously it would indiscriminately assign an expression to a LHS.
However, if there's a symbol in the expression its value might change
in the future, changing the expression the value had in the past,
most egregiously in cases like `x = x + 1`, which would become
equivalent to `x = (x + 1) + 1` etc.
